module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/**
+ * Copyright © Bold Brand Commerce Sp. z o.o. All rights reserved.
+ * See LICENSE.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Ergonode\Product\Infrastructure\Provider;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Ergonode\BatchAction\Domain\ValueObject\BatchActionFilter;
+use Ergonode\BatchAction\Infrastructure\Provider\FilteredQueryBuilderProviderInterface;
+use Ergonode\Core\Domain\ValueObject\Language;
+use Ergonode\Grid\FilterGridConfiguration;
+use Ergonode\Product\Infrastructure\Factory\DataSet\DbalQueryBuilderProductDataSetFactory;
+use Ergonode\Product\Infrastructure\Grid\ProductGridBuilder;
+
+class FilteredQueryBuilderProvider implements FilteredQueryBuilderProviderInterface
+{
+    private DbalQueryBuilderProductDataSetFactory $productQueryBuilderFactory;
+
+    private ProductGridBuilder $gridBuilder;
+
+    private ProductIdsProvider $productIdsProvider;
+
+    private Connection $connection;
+
+    public function __construct(
+        DbalQueryBuilderProductDataSetFactory $productQueryBuilderFactory,
+        ProductGridBuilder $gridBuilder,
+        ProductIdsProvider $productIdsProvider,
+        Connection $connection
+    ) {
+        $this->productQueryBuilderFactory = $productQueryBuilderFactory;
+        $this->gridBuilder = $gridBuilder;
+        $this->productIdsProvider = $productIdsProvider;
+        $this->connection = $connection;
+    }
+
+
+    public function provide(?BatchActionFilter $filter): QueryBuilder
+    {
+        $queryBuilder = $this->getQueryBuilder();
+        if ($filter) {
+            $filterQuery = $filter->getQuery();
+            if ($filterQuery) {
+                if ($filter->getIds()) {
+                    if ($filter->getIds()->isIncluded()) {
+                        //query and include id
+                        $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                        return $filteredQueryBuilder->orWhere($filteredQueryBuilder->expr()->in('id', ':productsIds'))
+                            ->setParameter(':productsIds', $filter->getIds()->getList(), Connection::PARAM_STR_ARRAY);
+                    }
+
+                    //query and exclude id
+                    $filteredQueryBuilder = $this->getFilteredQueryBuilder($filterQuery);
+
+                    return $filteredQueryBuilder->orWhere($filteredQueryBuilder->expr()->notIn('id', ':productsIds'))
```

Comment:
```suggestion
                    return $filteredQueryBuilder->andWhere($filteredQueryBuilder->expr()->notIn('id', ':productsIds'))
```
Example:
- [A,B,C] -> or not [B,D] => [A,B,C]  is not expected,
- [A,B,C] -> and not [B,D] => [A, C]  is expected,